File: java/src/java/com/d_project/qrcode/Polynomial.java
package com.d_project.qrcode;

/**
 * Polynomial
 * @author Kazuhiko Arase 
 */
class Polynomial {
	
	private final int[] num;

	public Polynomial(int[] num) {
		this(num, 0);
	}

	public Polynomial(int[] num, int shift) {

		int offset = 0;

		while (offset < num.length && num[offset] == 0) {
			offset++;
		}

		this.num = new int[num.length - offset + shift];
		System.arraycopy(num, offset, this.num, 0, num.length - offset);
	}

	public int get(int index) {
		return num[index];
	}

	public int getLength() {
		return num.length;
	}

	public String toString() {

		StringBuilder buffer = new StringBuilder();

		for (int i = 0; i < getLength(); i++) {
			if (i > 0) {
				buffer.append(",");
			}
			buffer.append(get(i) );
		}

		return buffer.toString();
	}
	
	public String toLogString() {

		StringBuilder buffer = new StringBuilder();

		for (int i = 0; i < getLength(); i++) {
			if (i > 0) {
				buffer.append(",");
			}
			buffer.append(QRMath.glog(get(i) ) );

		}

		return buffer.toString();
	}
	
	public Polynomial multiply(Polynomial e) {

		int[] num = new int[getLength() + e.getLength() - 1];

		for (int i = 0; i < getLength(); i++) {
			for (int j = 0; j < e.getLength(); j++) {
				num[i + j] ^= QRMath.gexp(QRMath.glog(get(i) ) + QRMath.glog(e.get(j) ) );
			}
		}

		return new Polynomial(num);
	}
	
	public Polynomial mod(Polynomial e) {

		if (getLength() - e.getLength() < 0) {
			return this;
		}

		// 最上位桁の比率
		int ratio = QRMath.glog(get(0) ) - QRMath.glog(e.get(0) );

		// コピー作成
		int[] num = new int[getLength()];
		for (int i = 0; i < getLength(); i++) {
			num[i] = get(i);
		}
		
		// 引き算して余りを計算
		for (int i = 0; i < e.getLength(); i++) {
			num[i] ^= QRMath.gexp(QRMath.glog(e.get(i) ) + ratio);
		}

		// 再帰計算
		return new Polynomial(num).mod(e);
	}
}

File: ts/src/ts/com/d_project/qrcode/Polynomial.ts
'use strict';
namespace com.d_project.qrcode {

  /**
   * Polynomial
   * @author Kazuhiko Arase
   */
  export class Polynomial {

    private num : number[];

    public constructor(num : number[], shift = 0) {

      var offset = 0;

      while (offset < num.length && num[offset] == 0) {
        offset += 1;
      }

      this.num = [];
      var len = num.length - offset;
      for (var i = 0; i < len; i += 1) {
        this.num.push(num[offset + i]);
      }
      for (var i = 0; i < shift; i += 1) {
        this.num.push(0);
      }
    }

    public getAt(index : number) : number{
      return this.num[index];
    }

    public getLength() : number {
      return this.num.length;
    }

    public toString() : string {
      var buffer = '';
      for (var i = 0; i < this.getLength(); i += 1) {
        if (i > 0) {
          buffer += ',';
        }
        buffer += this.getAt(i);
      }
      return buffer.toString();
    }

    public toLogString() : string {
      var buffer = '';
      for (var i = 0; i < this.getLength(); i += 1) {
        if (i > 0) {
          buffer += ',';
        }
        buffer += QRMath.glog(this.getAt(i) );
      }
      return buffer.toString();
    }

    public multiply(e : Polynomial) : Polynomial {
      var num : number[] = [];
      var len = this.getLength() + e.getLength() - 1;
      for (var i = 0; i < len; i += 1) {
        num.push(0);
      }
      for (var i = 0; i < this.getLength(); i += 1) {
        for (var j = 0; j < e.getLength(); j += 1) {
          num[i + j] ^= QRMath.gexp(QRMath.glog(this.getAt(i) ) +
            QRMath.glog(e.getAt(j) ) );
        }
      }
      return new Polynomial(num);
    }

    public mod(e : Polynomial) : Polynomial {

      if (this.getLength() - e.getLength() < 0) {
        return this;
      }

      var ratio = QRMath.glog(this.getAt(0) ) - QRMath.glog(e.getAt(0) );

      // create copy
      var num : number[] = [];
      for (var i = 0; i < this.getLength(); i += 1) {
        num.push(this.getAt(i) );
      }

      // subtract and calc rest.
      for (var i = 0; i < e.getLength(); i += 1) {
        num[i] ^= QRMath.gexp(QRMath.glog(e.getAt(i) ) + ratio);
      }

      // call recursively
      return new Polynomial(num).mod(e);
    }
  }
}
